pkg/mirroring/s3.go

```diff
@@ -0,0 +1,107 @@
+package mirroring
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/warptools/warpforge/wfapi"
+
+	"github.com/serum-errors/go-serum"
+)
+
+type S3Pusher struct {
+	client       *s3.Client
+	cfg          wfapi.S3PushConfig
+	existingKeys map[string]bool
+}
+
+func wareIdToKey(wareId wfapi.WareID) string {
+	return filepath.Join(wareId.Hash[0:3], wareId.Hash[3:6], wareId.Hash)
```

Comment:
As a TODO, we're doing this in a lot of places and maybe we should toss a default ware ID projection function somewhere.

Reply:
`wfapi.WareID.Subpath()` now does this.